invenio_communities/permissions.py

```diff
@@ -0,0 +1,465 @@
+    #     pass
+
+class AnyMember(Generator):
+    """Allows any member."""
+
+    def needs(self, record=None, **kwargs):
+        """Enabling Needs.
+
+        :param record: a member
+        """
+        member = record
+
+        # Symptom of how api.py objects are falsey if dict empty
+        if member is None:
+            return []
+
+        return [create_community_role_need(member.community_id, "member")]
+
+
+class IfCommunityRestricted(IfRestrictedBase):
+    """If Community is restricted."""
+
+    @classmethod
+    def get_visiblity(cls, record):
+        """Return visibility.
+
+        We have to make a DB call here to fetch the related community.
+
+        :param record: a member
+        """
+        member = record
+        community = Community.get_record(member.community_id)
+        return getattr(community.access, "visibility", None)
+
+    def __init__(self, then_, else_):
+        """Initialize."""
+        super().__init__(
+            IfCommunityRestricted.get_visiblity,
+            None,
+            "restricted",
+            "public",
+            then_,
+            else_,
+        )
+
+    def query_filter(self, **kwargs):
+        """Don't use this for queries.
+
+        See can_read_search_members.
+        """
+        raise NotImplementedError()
+
+
+# Permission Policy
+
+class CommunityPermissionPolicy(BasePermissionPolicy):
+    """Permissions for Community CRUD operations."""
+
+    # Community
+    can_create = [AuthenticatedUser(), SystemProcess()]
+
+    can_read = [
+        IfRestricted(
+            'visibility',
+            then_=[CommunityOwner()],
+            else_=[AnyUser()]),
+        ]
+
+    can_update = [CommunityOwner(), SystemProcess()]
+
+    can_delete = [CommunityOwner(), SystemProcess()]
+
+    can_search = [AnyUser(), SystemProcess()]
+
+    can_search_user_communities = [AuthenticatedUser(), SystemProcess()]
+
+    can_rename = [CommunityOwner(), SystemProcess()]
+
+    can_submit_record = [
+        IfPolicyClosed(
+            'record_policy',
+            then_=[CommunityOwner(), SystemProcess()],
+            else_=[IfRestricted(
+                'visibility',
+                then_=[CommunityOwner()],
+                else_=[AuthenticatedUser()]),
+            ],
+        ),
+    ]
+
+    # Placed here because passed record is a community
+    can_create_member = [CommunityOwner(), CommunityManager(), SystemProcess()]
+
+    can_search_members = [
+        SystemProcess(),
+        IfRestricted(
+            'visibility',
+            then_=[CommunityMember()],
+            else_=[AnyUser()]
+        ),
+    ]
+
+    # Members (passed record is a member)
+
+    # This is a new performance enhancing permission to be used when reading
+    # a record from the search results.
+    # Because can_search_members has already been applied, any user who got
+    # through can read the record.
+    can_read_search_members = [AnyUser(), SystemProcess()]
```

Comment:
This is the pattern I was mentioning. I don't know if it is applicable to other search scenarios. (I *think* it's applicable to the RequestTypes, but we will see...)